include/boost/geometry/algorithms/hausdorff_distance.hpp

```diff
@@ -0,0 +1,324 @@
+        
+        boost::geometry::detail::throw_on_empty_input(pnt);
+        boost::geometry::detail::throw_on_empty_input(rng);
+        
+        size_type  b = boost::size(rng);
+        result_type dis_min;
+        bool is_dis_min_set = false;
+        for(size_type i=0;i<b;i++)
+        {
+            result_type dis_temp = geometry::distance(pnt, range::at(rng, i), strategy);
+                if(!is_dis_min_set || dis_temp < dis_min)
+                {    
+                    dis_min=dis_temp;
+                    is_dis_min_set = true;
+                }
+        }
+        return dis_min;
+    }
+};
+#ifdef BOOST_GEOMETRY_ENABLE_SIMILARITY_RTREE
+struct range_range
+{
+    template <typename Range1, typename Range2, typename Strategy>
+    static inline typename distance_result
+        <
+            typename point_type<Range1>::type,
+            typename point_type<Range2>::type,
+            Strategy
+        >::type apply(Range1 const& r1, Range2 const& r2, Strategy const& strategy)
+    {
+        
+        typedef typename distance_result
+            <
+                typename point_type<Range1>::type,
+                typename point_type<Range2>::type,
+                Strategy
+            >::type result_type;
+
+        typedef typename boost::range_size<Range1>::type size_type;
+        typedef typename point_type<Range1>::type point_t;
+        
+        boost::geometry::detail::throw_on_empty_input(r1);
+        boost::geometry::detail::throw_on_empty_input(r2);
+        
+        size_type  a = boost::size(r1);
+        result_type dis_max=0;
+
+        //Computing the HausdorffDistance
+        typedef bgi::rtree<point_t, bgi::linear<4> > rtree_type;
+        rtree_type rtree(boost::begin(r2), boost::end(r2));
+        point_t res;
+        for(size_type i=0;i<a;i++)
+        {
+            size_type found= rtree.query(bgi::nearest(range::at(r1, i),1), &res);
+            result_type dis_min=geometry::distance(range::at(r1,i),res);
+            if (dis_min > dis_max )
+            {
+              dis_max = dis_min;
+            }
+        }
+        return dis_max;
+    }
+};
+#else
+struct range_range
+{
+    template <typename Range1, typename Range2, typename Strategy>
+    static inline typename distance_result
+        <
+            typename point_type<Range1>::type,
+            typename point_type<Range2>::type,
+            Strategy
+        >::type apply(Range1 const& r1, Range2 const& r2, Strategy const& strategy)
+    {
+        
+        typedef typename distance_result
+            <
+                typename point_type<Range1>::type,
+                typename point_type<Range2>::type,
+                Strategy
+            >::type result_type;
+
+        typedef typename boost::range_size<Range1>::type size_type;
+        
+        boost::geometry::detail::throw_on_empty_input(r1);
+        boost::geometry::detail::throw_on_empty_input(r2);
```

Comment:
Note that everything above from this point in the struct is the same for rtree and loop variants. To not duplicate the code you could enable only the code below with `#ifdef BOOST_GEOMETRY_ENABLE_SIMILARITY_RTREE` not the whole struct.

Even more flexible would be implementing the loop and rtree search in a separate struct and take it as Policy (`range_range` template argument defining `apply()` static member function) and pick this Policy later in `dispatch::hausdorff_distance` definition for two linestrings.

Reply:
And then you could call the Policy with a loop in `point_range` directly. This would also fix the problem that `point_range` checks the size of a linestring each time it's called and this is not needed since it's already checked here.

Reply:
merged the previous 2 struct ( range_range with rtree and with loop ) in single struct using       
`#ifdef BOOST_GEOMETRY_ENABLE_SIMILARITY_RTREE` condition.